### PR TITLE
doc: gsg: Edit instructions to set SDK env vars

### PR DESCRIPTION
**Pull request by greg-fer**
Edited the changes added in PR #46319.
The changes aim at including the variable info
in each OS installation procedure for major visibility.
Also added subsections on zephyr_sdk.rst page.

Signed-off-by: Grzegorz Ferenc <Grzegorz.Ferenc@nordicsemi.no>

### CODE REVIEW

File: doc/develop/getting_started/index.rst
Comment (stephanosio) [inline doc/develop/getting_started/index.rst:538]: See https://github.com/zephyrproject-rtos/zephyr/pull/46319#discussion_r890599871 and https://github.com/zephyrproject-rtos/zephyr/pull/46319#discussion_r890602240

File: doc/develop/getting_started/index.rst
Comment (stephanosio) [inline doc/develop/getting_started/index.rst:589]: same deal

File: doc/develop/getting_started/index.rst
Comment (stephanosio) [inline doc/develop/getting_started/index.rst:634]: ditto

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (stephanosio) [inline doc/develop/toolchains/zephyr_sdk.rst:41]: I do not see how this is an improvement over the existing documentation.

It would indeed be useful to have an "environment variable glossary" page under `doc/develop/toolchains` that lists all the toolchain-related environment variables (referenced by `:envvar:`) with a detailed description, but what is done here seems like a half-baked solution.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst:589]: Edited as per comment https://github.com/zephyrproject-rtos/zephyr/pull/46365#discussion_r892052828

File: doc/develop/getting_started/index.rst
Comment (mbolivar-nordic) [inline doc/develop/getting_started/index.rst:540]: Nothing wrong with this line, but please refer to the commit (0a758c36c864cc138941848f42c8ea4aa7dc95b3) instead of the pull request number (PR#46319) in the commit log.

Git is a decentralized version control system. Commits are "forever" and "global" objects across forks. Pull request numbers are specific to individual forks. So the log will be clearer if you use the commit instead. This also allows people to figure out the changes you are talking about even when they are offline.

File: doc/develop/getting_started/index.rst
Comment (mbolivar-nordic) [inline doc/develop/getting_started/index.rst:537]: Wait, why? These environment variables are optional these days; the default toolchain variant is `zephyr` if not set, and the installation directory is auto-discovered from a set of standard paths if not set. The "Make sure" language here suggests they are mandatory.

Comment (glarsennordic) [reply doc/develop/getting_started/index.rst:537]: I like this edit, because for some projects, the `ZEPHYR_TOOLCHAIN_VARIANT` value was set to something other than `zephyr`, and when installing the new Zephyr toolchain, the old value is left untouched. This means the new install actually doesn't use the new toolchain, but uses whatever toolchain was previously installed. 

I think it is useful to have this mentioned as part of the gsg, because this is (indirectly) where a lot of users will be going to update their toolchain, and they (like me, at first) may not know or remember to even check the `ZEPHYR_TOOLCHAIN_VARIANT`

I agree, however, that the "Make sure" phrasing is misleading.

Perhaps better phrasing would be:

```rst
.. note::
        Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
        If you are installing the Zephyr toolchain to replace an existing Zephyr-compatible toolchain, check
        whether the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
        are set.
	
        If they are not set, you may proceed without making any changes (the new Zephyr toolchain will be selected
        by default).

        If they have a value, this will override Zephyr's default toolchain selection, and may prevent the Zephyr
        toolchain from being selected. In this case, either delete them, or set :envvar:`ZEPHYR_TOOLCHAIN_VARIANT`
	to ``zephyr`` and set :envvar:`ZEPHYR_SDK_INSTALL_DIR` to your Zephyr SDK installation directory. This
	will cause the Zephyr toolchain to be selected.
```

Though it is unclear to me what the ``zephyr`` toolchain variant actually does (does it cause the default selection process to occur, or does it specifically choose the Zephyr toolchain? Is "the Zephyr toolchain" even correct terminology, here?

Comment (greg-fer) [reply doc/develop/getting_started/index.rst:537]: @glarsennordic , applied an edited version of your proposal *before* the installation steps. This is because the proposal reads like something you should be aware of before you start installing the SDK. Placing it before also allows to avoid duplicating the content in each tab.

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst]: This is now applied, thanks!

**Review by stephanosio — DISMISSED** [review]

**Review by glarsennordic — CHANGES_REQUESTED** [review]

**Comment by glarsennordic** [comment]
Looks great -- I'm pretty sure my request for changes was not official since I am not a reviewer, but ping me if I am still somehow blocking merge

**Comment by glarsennordic** [comment]
@stephanosio It looks like my change requests may have addressed your requested changes as well

Do you find the latest version satisfactory?

File: doc/develop/getting_started/index.rst
Comment (tejlmand) [inline doc/develop/getting_started/index.rst:492]: I don't think this belongs here.

New users starting out with Zephyr or installing on a clean machine doesn't need to carry out those steps. They are imho just noise in those cases.

We should try to keep the guide as simple as possible without unnecessary steps for newcomers.

Instead this belongs in an *Updating Zephyr SDK* section or a trouble shooting section as this may only happen to users updating an existing system (which is beyond getting started).

If you add a *Update Zephyr SDK* section to the [Zephyr SDK](https://docs.zephyrproject.org/latest/develop/toolchains/zephyr_sdk.html) page then you can simple change the note to:
```suggestion
.. note::
   If you're updating the Zephyr SDK from an older release, please read :ref:`update_sdk`
``` 

Another option is to change the note to pure text and move it downwards to a *Troubleshooting* subsection. That subsection can then include the [asking-for-help](https://docs.zephyrproject.org/latest/develop/getting_started/index.html#asking-for-help) subsection.

This will be more aligned with what we have for [devicetree](https://docs.zephyrproject.org/latest/build/dts/index.html) and [west](https://docs.zephyrproject.org/latest/develop/west/index.html) where there exists dedicated [Troubleshooting](https://docs.zephyrproject.org/latest/develop/west/troubleshooting.html) sections for users experiencing problems.

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (tejlmand) [inline doc/develop/toolchains/zephyr_sdk.rst:37]: the build system doesn't assume anything.
It defaults to look for the Zephyr SDK if no toolchain has been selected.

How about:
```suggestion
If no toolchain is selected then the build system will look for Zephyr SDK and use a toolcahin from there.
```

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (tejlmand) [inline doc/develop/toolchains/zephyr_sdk.rst:43]: > you must register

is this a **must** ?
And since when ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this.

I acknowledge that we can improve troubleshooting of Zephyr SDK updates, but I believe the **note** is misplaced, and I propose moving it to other location.

**Comment by tejlmand** [comment]
@glarsennordic if you're satisfied with the latest changes, please approve the PR.

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst:492]: I went for adding a Troubleshooting section.

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (greg-fer) [reply doc/develop/toolchains/zephyr_sdk.rst:43]: @carlescufi , this is a question for you :)

**Comment by greg-fer** [comment]
Addressed some of your comments, @tejlmand , although I was not sure where to best place the Troubleshooting section.
One comment is pending answer from @carlescufi , as I've just restructured what he wrote.

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (carlescufi) [reply doc/develop/toolchains/zephyr_sdk.rst:42]: It´s a must if you do not use environment variables @tejlmand?

In order for the SDK to be found by the build system you need:
- To install it in a default location OR
- to register it with CMake OR
- to set the `ZEPHYR_SDK_INSTALL_DIR`

isn't this correct?

Comment (tejlmand) [reply doc/develop/toolchains/zephyr_sdk.rst:42]: As long as you don't state how `ZEPHYR_SDK_INSTALL_DIR` is set.
It is correct that it must be set if not installed in default search path or registered as a CMake package.

But it doesn't have to be set in environment. It may be set in the CMake cache using `-D`.
Another option is to use the [Zephyr Build Configuration CMake package](https://docs.zephyrproject.org/latest/build/zephyr_cmake_package.html#zephyr-build-configuration-cmake-package), which can be used on project level to ensure a given toolchain is used. For example this is what is done by NCS and the NCS Toolchain.
Using a build configuration CMake package allows a project to define a specific toolchain and remove the need for user interaction.

Note: I just discovered that @greg-fer only reformatted the text and actually didn't change the wording.

How about this:

> If you install the Zephyr SDK outside any of the default locations (listed in the operating system specific instructions below) and you want automatic discovery of the Zephyr SDK, then you must register the Zephyr SDK in the CMake package registry by running the setup script.
> If you decide not to register the Zephyr SDK in the CMake registry, then the :envvar:`ZEPHYR_SDK_INSTALL_DIR` can be used to point to the Zephyr SDK installation directory.

That way we don't state that you must use one of those two ways, but that you can use one of those.
And at this point I don't believe we need to mention all possible options, as that should be generally describe, see https://github.com/zephyrproject-rtos/zephyr/issues/40389

File: doc/develop/getting_started/index.rst
Comment (tejlmand) [inline doc/develop/getting_started/index.rst:724]: > corresponding toolchain will be selected at build time instead of the Zephyr SDK.

Well, only if the value is different from `zephyr` :wink: 

How about:
```suggestion
If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` is set, the corresponding toolchain will be selected at build time. Zephyr SDK is identified by the value `zephyr`.
If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` environment variable is not ``zephyr``, then either unset it or change its value to ``zephyr`` to make sure the Zephyr SDK is selected.
```

Note: this still only refers to the environment variable. In pratice the CMakeCache takes precedence, but I think we can wait for #40389 before cleaning up this.
For now, i'm satisfied with just referring to the envvar to keep things simple for the users.

File: doc/develop/getting_started/index.rst
Comment (tejlmand) [inline doc/develop/getting_started/index.rst:727]: Not sure I understand the meaning of this sentence:
>  If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location for the selected toolchain. 

So if I set `ZEPHYR_TOOLCHAIN_VARIANT` to `gnuarmemb`, then how will setting `ZEPHYR_SDK_INSTALL_DIR` impact the lookup of the `gnuarmemb` toolchain ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for creating a trouble shooting section.

Did a closer review at the text itself for that section.

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst:727]: Summoning @glarsennordic  for this technical question :)

File: doc/develop/toolchains/zephyr_sdk.rst
Comment (greg-fer) [reply doc/develop/toolchains/zephyr_sdk.rst:42]: Applied suggestion...

**Comment by greg-fer** [comment]
Applied some of your edits, @tejlmand , @carlescufi in a separate commit. I'll squash it after we nail the wording.

**Review by glarsennordic — DISMISSED** [review]

File: doc/develop/getting_started/index.rst
Comment (glarsennordic) [reply doc/develop/getting_started/index.rst:728]: My apologies, I'm not actually 100% sure whether `ZEPHYR_SDK_INSTALL_DIR` applies only when `ZEPHYR_TOOLCHAIN_VARIANT="zephyr"` or if it impacts the lookup location for all of the toolchains.

> the selected toolchain

was sloppy phrasing on my part. It may be accurate, or it may not be. Someone actually familiar with the build configuration variables needs to chime in, and I don't know who that is.

If it only affects the zephyr SDK specifically, than better phrasing would be:

```rst
* If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location
  for the Zephyr SDK. If you install the Zephyr SDK to one of the locations recommended below,
```

Do we have any kind of reference documentation specifically dedicated to these environment variables, their function, and their possible values? That would help quite a bit.

**Comment by glarsennordic** [comment]
Approving with regard to the section in the gsg itself, with the major caveat that the statement about what `ZEPHYR_SDK_INSTALL_DIR` does may well be inaccurate. Someone familiar with these environment variables should review the final version of this section.

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst]: Applied edits.

**Comment by greg-fer** [comment]
@glarsennordic , @tejlmand , @carlescufi , please have another look.

- Modified "the selected toolchain" sentence as per feedback from Torsten and Georges.
- Moved the section to "Beyong GSG" page, as writing about updating made more sense there. I kept however the troubleshooting section with the section. Granted, there is some duplication, but I didn't think removing the info from GSG entirely was OK.
- Added links to "Important environmental variables" section, as per Georges' feedback.

File: doc/develop/getting_started/index.rst
Comment (mbolivar-nordic) [reply doc/develop/getting_started/index.rst]: @greg-fer the commit log says:

```
Edited the changes added in commit #0a758c3.
```

it should say:

```
Edited the changes added in commit 0a758c3.
```

Commit SHAs are not numbers :).

File: doc/develop/getting_started/index.rst
Comment (mbolivar-nordic) [reply doc/develop/getting_started/index.rst]: @glarsennordic 

The text

```
        If you are installing the Zephyr toolchain to replace an existing Zephyr-compatible toolchain, check
```

seems out of place for a document called the "Getting started guide".

A perennial problem that we face with this document is that everybody wants to include the little bits and pieces that they learned over time when adjusting their environment to fit their use cases, but these end up being noise for people who are really just getting started for the first time. We need to keep fighting to keep this document as short as possible, because installing zephyr is already complex enough.

File: doc/develop/env_vars.rst
Comment (mbolivar-nordic) [inline doc/develop/env_vars.rst:190]:
```suggestion
You might need to update some of these variables when you :ref:`update the Zephyr SDK toolchain <gs_toolchain_update>`.
```

Similar issue throughout -- shouldn't we say "the" Zephyr SDK toolchain, with a definite article?

File: doc/develop/getting_started/index.rst
Comment (mbolivar-nordic) [inline doc/develop/getting_started/index.rst:725]: Why is this needed?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Generally I am fine with adding a troubleshooting section.

However, this now seems to be making "Asking for help" a subsection of "troubleshooting". The commit message provides no justification as to why. My first instinct is to disagree with this change, as the two are different. I might not have any trouble to shoot at all with the basic getting started guide, but I would still want to know where to ask for help as I move along learning more about zephyr and running into problems with more advanced areas.

File: doc/develop/getting_started/index.rst
Comment (greg-fer) [reply doc/develop/getting_started/index.rst:725]: OK, reverting.

**Comment by greg-fer** [comment]
@mbolivar-nordic , thanks for the suggestions! Applied all of them, and I agree about the "Ask for Help" section.

**Review by tejlmand — DISMISSED** [review]
lgtm.

File: doc/develop/getting_started/index.rst
Comment (tejlmand) [reply doc/develop/getting_started/index.rst]: I just discovered `toolchain` has been changed to `Zephyr SDK` making the sentence clearer in meaning 👍 .
No further comments.

**Comment by greg-fer** [comment]
@mbolivar-nordic , amended the commit description for hash "number". Please check the PR again if you have time.

**Comment by greg-fer** [comment]
@tejlmand , looks like your approval has been dismissed with my push that amended the commit description. Could you approve again?

@glarsennordic , your check and approval would be also appreciated.

**Review by carlescufi — DISMISSED** [review]
This looks good to me now, thanks @greg-fer!

File: doc/develop/beyond-GSG.rst
Comment (stephanosio) [inline doc/develop/beyond-GSG.rst:93]: > one of the locations recommended below

there aren't any locations listed **below**.

**Review by stephanosio — CHANGES_REQUESTED** [review]
In general, please revise the formatting such that there is no more than 100 characters per line.

**Review by stephanosio — DISMISSED** [review]

File: doc/develop/getting_started/index.rst
Comment (tejlmand) [inline doc/develop/getting_started/index.rst:721]:
```suggestion
For more information about these environment variables in Zephyr, see :ref:`env_vars_important`.
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm

**Review by stephanosio — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
The language is a bit inconsistent between "the Zephyr SDK" (which I think is right) and just plain "Zephyr SDK", but that's minor. Everything else LGTM.

**Pull request merged**
